Objective-C/CBLProperties.mm

```diff
@@ -22,85 +22,84 @@
+    _properties = [self convertProperties: properties];
+    
+    _changedNames = nil;
+    [self markChangedKeys: [_properties allKeys]];
 }
 
 
-static NSNumber* numberProperty(NSDictionary* dict, NSString* key) {
-    id obj = dict[key];
-    return [obj isKindOfClass: [NSNumber class]] ? obj : nil;
+static NSNumber* numberProperty(id value) {
+    return [value isKindOfClass: [NSNumber class]] ? value : nil;
 }
 
 
 - (BOOL) booleanForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return [numberProperty(_properties ? _properties : _rootProps, key) boolValue];
-    else
-        return FLValue_AsBool([self fleeceValueForKey: key]);
-    return NO;
+    id v = _properties ? _properties[key] : nil;
+    return v ? [numberProperty(v) boolValue] : FLValue_AsBool([self fleeceValueForKey: key]);
 }
 
 
 - (nullable NSDate*) dateForKey: (NSString*)key {
     NSString* dateStr = [self stringForKey: key];
-    return [CBLJSON dateWithJSONObject: dateStr];
+    return dateStr ? [CBLJSON dateWithJSONObject: dateStr] : nil;
 }
 
 
 - (double) doubleForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return [numberProperty(_properties ? _properties : _rootProps, key) doubleValue];
-    else
-        return FLValue_AsDouble([self fleeceValueForKey: key]);
+    id v = _properties ? _properties[key] : nil;
+    return v ? [numberProperty(v) doubleValue] : FLValue_AsDouble([self fleeceValueForKey: key]);
 }
 
 
 - (float) floatForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return [numberProperty(_properties ? _properties : _rootProps, key) floatValue];
-    else
-        return FLValue_AsFloat([self fleeceValueForKey: key]);
+    id v = _properties ? _properties[key] : nil;
+    return v ? [numberProperty(v) floatValue] : FLValue_AsFloat([self fleeceValueForKey: key]);
 }
 
 
 - (NSInteger) integerForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return [numberProperty(_properties ? _properties : _rootProps, key) integerValue];
-    else
-        return (NSInteger)FLValue_AsInt([self fleeceValueForKey: key]);
+    id v = _properties ? _properties[key] : nil;
+    return v ? [numberProperty(v) integerValue] : FLValue_AsInt([self fleeceValueForKey: key]);
 }
 
 
 - (nullable id) objectForKey: (NSString*)key {
-    return [self objectForKeyedSubscript: key];
+    return self[key];
 }
 
 
 - (nullable NSString*) stringForKey: (NSString*)key {
-    if (_properties || _rootProps)
-        return _properties ? _properties[key] : _rootProps[key];
-    else
-        return slice2string((FLValue_AsString([self fleeceValueForKey: key])));
+    id v = _properties ? _properties[key] : nil;
+    return v ? $castIf(NSString, v) : slice2string((FLValue_AsString([self fleeceValueForKey: key])));
+}
+
+
+- (CBLSubdocument*) subdocumentForKey: (NSString*)key {
+    id subdoc = self[key];
+    if (!subdoc) {
+        subdoc = [self subdocumentWithRoot: nullptr withProperties: nil forKey: key];
```

Comment:
By the API design, -subdocumentForKey: has the same behavior as Database's -documentWithID:, which will do get-or-create. We could discuss about next week. I will setup a meeting.